cl/alerts/tasks.py

```diff
@@ -62,20 +83,54 @@ def send_docket_alert(d_pk, since):
             date_created__gte=since, docket=docket
         )
 
+        trigger = False
         if new_des.count() > 0:
+            trigger = True
+        elif (
+            settings.PACER_USERNAME
+        ):  # if no credentials available, scraping is pointless. Required to pass tests.
+            # No new docket entries, check date_last_filing
+            if (
+                not docket.date_last_filing
+                or docket.date_last_filing < datetime.today().date()
+                and docket.date_last_filing < since.date()
+            ):  # Scrape iquery if date_last_filing is before today and before the last trigger
+                update_docket_date_last_filing(docket)
+            if docket.date_last_filing > since.date():
+                trigger = True
```

Comment:
I'd still like to pull this logic out of this task and give it its own task. I think it's a lot tidier to have one task that gets stuff from the network, and another that sends emails, rather than one that sometimes does both.

Do you think you could do that, and then use a Celery chain in your management command to do something like:

```
chain(
    crawl_iquery_page.s(docket), 
    send_alerts.s(), 
    add_or_update_recap_docket.s()
).apply_async()
```

(docs for chains are here: https://docs.celeryproject.org/en/latest/userguide/canvas.html#chains)

That'd also update the search index with the new values, which is a generally good idea as we're doing this kind of thing. 

Thoughts?

Reply:
Ok, I'm pulling it out. Since it's now only being run on a specific schedule, we don't actually need any of this logic - I can have the daemon call the `update_docket_info_iqeury` function directly

Reply:
Can't quite figure out the chains - do I need to make send_docket_alert return data in the format add_or_update_recap_docket expects?

Reply:
I think I'm going to skip the chains and do it manually, put everything in a single function that gets called as a task.

Reply:
Sure. That should work too. The only thing to watch out for in that case is that you don't want to call asynchronous tasks from within async tasks, since it can create deadlocks:

1. Async task number 1 starts
2. Async task number 1 calls task number 2 asynchronously
3. Task two crashes
4. Task one waits for it to return
5. The worker that was running task one never gets freed.
6. This repeats a bunch of times until no more workers remain.

Happily, tasks are just functions, so you can definitely do what you suggest. Just don't do async inside async. (Sorry if you already know all this stuff; Celery has some nasty footguns.)

Reply:
Getting a weird error when calling add_or_update_recap_docket during a test that doesn't show up when running ` docker exec -it cl-django python /opt/courtlistener/manage.py pacer_iquery_scraper` directly. 

`docker exec -it cl-django python /opt/courtlistener/manage.py test --keepdb cl.alerts.tests.DocketAlertTest`

```
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/subj_alt_name.py:22: UserWarning: Error importing pyasn1, subjectAltName check for SSL peer verification will be disabled.  Import error is: No module named pyasn1.type
  warnings.warn(import_error_msg)
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/ssl_peer_verification.py:25: UserWarning: SubjectAltName support is disabled - check pyasn1 package installation to enable
  warnings.warn(SUBJ_ALT_NAME_SUPPORT_MSG)
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/subj_alt_name.py:22: UserWarning: Error importing pyasn1, subjectAltName check for SSL peer verification will be disabled.  Import error is: No module named pyasn1.type
  warnings.warn(import_error_msg)

Warning: No such file or directory: /var/log/juriscraper/debug.log. Have you created the directory for the log?
Juriscraper will continue to run, and all logs will be sent to stderr.
Using existing test database for alias 'default'...
Cache table 'django_cache' already exists.
System check identified no issues (0 silenced).
.2020-05-12 13:51:21,273 - INFO: Running case query for case ID '191424'
INFO:Logger:Running case query for case ID '191424'
/usr/local/lib/python2.7/site-packages/html5lib/_ihatexml.py:257: DataLossWarning: Coercing non-XML name
  warnings.warn("Coercing non-XML name", DataLossWarning)
E.2020-05-12 13:51:24,783 - INFO: Running case query for case ID '191424'
INFO:Logger:Running case query for case ID '191424'
E
======================================================================
ERROR: test_nothing_happens_for_timers_after_pacer_docket_date (cl.alerts.tests.DocketAlertTest)
Do we avoid sending alerts for timers after pacer date_last_filing?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/courtlistener/cl/alerts/tests.py", line 147, in test_nothing_happens_for_timers_after_pacer_docket_date
    update_docket_and_send_alert(self.docket2, self.new)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 169, in update_docket_and_send_alert
    update_docket_info_iqeury(docket)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 80, in update_docket_info_iqeury
    {"docket_pk": docket.pk, "content_updated": changes}
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/search/tasks.py", line 97, in add_or_update_recap_docket
    si.add(d.as_search_list())
  File "/opt/courtlistener/cl/search/models.py", line 753, in as_search_list
    "court": self.court.full_name,
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 184, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 159, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
DoesNotExist: Court matching query does not exist.

======================================================================
ERROR: test_triggering_pacer_docket_alert (cl.alerts.tests.DocketAlertTest)
Does the alert trigger for pacer date_last_filing?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/courtlistener/cl/alerts/tests.py", line 138, in test_triggering_pacer_docket_alert
    update_docket_and_send_alert(self.docket2, self.old)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 169, in update_docket_and_send_alert
    update_docket_info_iqeury(docket)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 80, in update_docket_info_iqeury
    {"docket_pk": docket.pk, "content_updated": changes}
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/search/tasks.py", line 97, in add_or_update_recap_docket
    si.add(d.as_search_list())
  File "/opt/courtlistener/cl/search/models.py", line 753, in as_search_list
    "court": self.court.full_name,
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 184, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 159, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
DoesNotExist: Court matching query does not exist.

----------------------------------------------------------------------
Ran 4 tests in 7.200s

FAILED (errors=2)
Preserving test database for alias 'default'...
```

I'm guessing the test database isn't set up to be able to run this. Should I just get rid of the tests? For now I added a hack to skip this line during tests.